Patch GRPC upstream to use only non-deprecated OpenSSL3 APIs.
Index: grpc/test/core/handshake/client_ssl.cc
===================================================================
--- grpc.orig/test/core/handshake/client_ssl.cc
+++ grpc/test/core/handshake/client_ssl.cc
@@ -193,11 +193,10 @@ static void ssl_server_info_callback(con
 static void server_thread(void* arg) {
   const server_args* args = static_cast<server_args*>(arg);
 
-  SSL_load_error_strings();
-  OpenSSL_add_ssl_algorithms();
+  OPENSSL_init_ssl(0, NULL);
   args->ssl_library_info->Notify();
 
-  const SSL_METHOD* method = TLSv1_2_server_method();
+  const SSL_METHOD* method = TLS_server_method();
   SSL_CTX* ctx = SSL_CTX_new(method);
   if (!ctx) {
     perror("Unable to create SSL context");
@@ -232,13 +231,6 @@ static void server_thread(void* arg) {
     grpc_core::Crash("Couldn't set server cipher list.");
   }
 
-  // Enable automatic curve selection. This is a NO-OP when using OpenSSL
-  // versions > 1.0.2.
-  if (!SSL_CTX_set_ecdh_auto(ctx, /*onoff=*/1)) {
-    ERR_print_errors_fp(stderr);
-    grpc_core::Crash("Couldn't set automatic curve selection.");
-  }
-
   // Register the ALPN selection callback.
   SSL_CTX_set_alpn_select_cb(ctx, alpn_select_cb, args->alpn_preferred);
 
@@ -386,8 +378,6 @@ TEST(ClientSslTest, MainTest) {
   // and sanity checks the client_ssl_test.
   ASSERT_FALSE(client_ssl_test(const_cast<char*>("foo")));
 #endif  // OPENSSL_IS_BORING_SSL
-  // Clean up the SSL libraries.
-  EVP_cleanup();
 }
 
 #endif  // GRPC_POSIX_SOCKET_TCP
Index: grpc/test/core/handshake/server_ssl_common.cc
===================================================================
--- grpc.orig/test/core/handshake/server_ssl_common.cc
+++ grpc/test/core/handshake/server_ssl_common.cc
@@ -187,7 +187,7 @@ bool server_ssl_test(const char* alpn_li
   // server port.
   s.Await();
 
-  const SSL_METHOD* method = TLSv1_2_client_method();
+  const SSL_METHOD* method = TLS_client_method();
   SSL_CTX* ctx = SSL_CTX_new(method);
   if (!ctx) {
     perror("Unable to create SSL context");
@@ -280,4 +280,4 @@ bool server_ssl_test(const char* alpn_li
   return success;
 }
 
-void CleanupSslLibrary() { EVP_cleanup(); }
+void CleanupSslLibrary() { }
Index: grpc/src/core/tsi/ssl_transport_security.cc
===================================================================
--- grpc.orig/src/core/tsi/ssl_transport_security.cc
+++ grpc/src/core/tsi/ssl_transport_security.cc
@@ -1636,7 +1636,7 @@ static tsi_result ssl_handshaker_result_
   unsigned int alpn_selected_len;
   const tsi_ssl_handshaker_result* impl =
       reinterpret_cast<const tsi_ssl_handshaker_result*>(self);
-  X509* peer_cert = SSL_get_peer_certificate(impl->ssl);
+  X509* peer_cert = SSL_get1_peer_certificate(impl->ssl);
   if (peer_cert != nullptr) {
     result = peer_from_x509(peer_cert, 1, peer);
     X509_free(peer_cert);
